frame/grandpa/src/equivocation.rs

```diff
@@ -0,0 +1,406 @@
+//! reporting extrinsics (from the client that's running the GRANDPA protocol).
+//! And in a runtime context, so that the GRANDPA module can validate the
+//! equivocation proofs in the extrinsic and report the offences.
+//!
+
+use sp_std::prelude::*;
+
+use codec::{self as codec, Decode, Encode};
+use frame_support::{debug, dispatch::IsSubType, traits::KeyOwnerProofSystem};
+use frame_system::offchain::{AppCrypto, CreateSignedTransaction, Signer};
+use sp_finality_grandpa::{EquivocationProof, RoundNumber, SetId};
+use sp_runtime::{
+	traits::{DispatchInfoOf, SignedExtension},
+	transaction_validity::{
+		InvalidTransaction, TransactionValidity, TransactionValidityError, ValidTransaction,
+	},
+	DispatchResult, Perbill,
+};
+use sp_staking::{
+	offence::{Kind, Offence, OffenceError, ReportOffence},
+	SessionIndex,
+};
+
+/// Ensure that equivocation reports are only processed if valid.
+#[derive(Encode, Decode, Clone, Eq, PartialEq)]
+pub struct ValidateEquivocationReport<T>(sp_std::marker::PhantomData<T>);
+
+impl<T> Default for ValidateEquivocationReport<T> {
+	fn default() -> ValidateEquivocationReport<T> {
+		ValidateEquivocationReport::new()
+	}
+}
+
+impl<T> ValidateEquivocationReport<T> {
+	pub fn new() -> ValidateEquivocationReport<T> {
+		ValidateEquivocationReport(Default::default())
+	}
+}
+
+impl<T> sp_std::fmt::Debug for ValidateEquivocationReport<T> {
+	fn fmt(&self, f: &mut sp_std::fmt::Formatter) -> sp_std::fmt::Result {
+		write!(f, "ValidateEquivocationReport<T>")
+	}
+}
+
+/// Custom validity error used when validating equivocation reports.
+#[derive(Debug)]
+#[repr(u8)]
+pub enum ReportEquivocationValidityError {
+	/// The proof provided in the report is not valid.
+	InvalidEquivocationProof = 1,
+	/// The proof provided in the report is not valid.
+	InvalidKeyOwnershipProof = 2,
+	/// The set id provided in the report is not valid.
+	InvalidSetId = 3,
+	/// The session index provided in the report is not valid.
+	InvalidSession = 4,
+}
+
+impl From<ReportEquivocationValidityError> for TransactionValidityError {
+	fn from(e: ReportEquivocationValidityError) -> TransactionValidityError {
+		TransactionValidityError::from(InvalidTransaction::Custom(e as u8))
+	}
+}
+
+impl<T: super::Trait + Send + Sync> SignedExtension for ValidateEquivocationReport<T>
+where
+	<T as frame_system::Trait>::Call: IsSubType<super::Module<T>, T>,
+{
+	const IDENTIFIER: &'static str = "ValidateEquivocationReport";
+	type AccountId = T::AccountId;
+	type Call = <T as frame_system::Trait>::Call;
+	type AdditionalSigned = ();
+	type Pre = ();
+
+	fn additional_signed(
+		&self,
+	) -> sp_std::result::Result<Self::AdditionalSigned, TransactionValidityError> {
+		Ok(())
+	}
+
+	fn validate(
+		&self,
+		_who: &Self::AccountId,
+		call: &Self::Call,
+		_info: &DispatchInfoOf<Self::Call>,
+		_len: usize,
+	) -> TransactionValidity {
+		let (equivocation_proof, key_owner_proof) = match call.is_sub_type() {
```

Comment:
Initially this was checked in the dispatchable and there was no signed extension. We want to avoid users being able to spam the chain with fake equivocation reports, one way to do it is to have a proper weight here and make sure that users pay for the transaction (this was what I did initially). But this requires that the reporting account has enough balance to pay these fees. So instead we decided to make the transaction free and validation happens in the signed extension, if the signed extension is not added then this doesn't work properly.